packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1933,12 +2011,34 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
 
   @override
   void didChangeNext(Route<dynamic>? nextRoute) {
+    // In some cases, we need to get the delegated transition from a route before
+    // it is built, but some routes need to derive a delegated transition from
+    // the context. So if the next route does not have a `BuildContext` or a
+    // delegatedTransition, call nextRoute.setDelegatedTransition using the
+    // current `BuildContext`.
+    if (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition == null && nextRoute.subtreeContext == null && this.subtreeContext != null) {
+      nextRoute.setDelegatedTransition(this.subtreeContext!);
```

Comment:
I imagine it will be the sub class to set the delegate directly in their didChangePrevious? also is the reason that we have this method is that we the route don't have access to context? if that is the case maybe the Route base class or the modalRoute provide a method like didChangeDependencies.